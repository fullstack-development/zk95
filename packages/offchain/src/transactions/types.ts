import { Script, Unit } from 'lucid-cardano';

export type PoolInfo = {
  txHash: string;
  nominal: number;
  treeTokenUnit: Unit;
  vaultTokenUnit: Unit;
  zeroValue: string;
  treeHeight: number;
  mixerScript: Script;
};
