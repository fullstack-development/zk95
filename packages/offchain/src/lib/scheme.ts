import { Data } from 'lucid-cardano';

const MerkleTreeState = Data.Object({
  nextLeaf: Data.Integer(),
  tree: Data.Any(),
});

const DepositDatum = Data.Object({
  merkleTreeState: MerkleTreeState,
  merkleTreeRoot: Data.Nullable(Data.Integer()),
});

export const MixerDatum = Data.Enum([
  Data.Object({ DepositTree: Data.Object({ DepositDatum }) }),
  Data.Literal('Vault'),
]);

export const MixerRedeemer = Data.Enum([
  Data.Object({ Deposit: Data.Tuple([Data.Bytes()]) }),
  Data.Literal('Topup'),
  Data.Literal('Withdraw'),
]);

export type MixerRedeemer = Data.Static<typeof MixerRedeemer>;

export type MixerDatum = Data.Static<typeof MixerDatum>;
