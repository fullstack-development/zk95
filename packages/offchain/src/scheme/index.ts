import { Data } from 'lucid-cardano';

export const MerkleTreeConfig = Data.Object({
  height: Data.Integer(),
  zeroHash: Data.Bytes(),
});

export type MixerConfig = Data.Static<typeof MixerConfig>;
export const MixerConfig = Data.Object({
  policyId: Data.Bytes(),
  treeTokenName: Data.Bytes(),
  vaultTokenName: Data.Bytes(),
  nullifiers_token_name: Data.Bytes(),
  poolNominal: Data.Integer(),
  relayerPkh: Data.Bytes(),
  merkleTreeConfig: MerkleTreeConfig,
});

export type MerkleTree = Data.Static<typeof MerkleTree>;
export const MerkleTree = Data.Object({
  root: Data.Bytes(),
  leafs: Data.Array(Data.Bytes()),
});

export type MixerDatum = Data.Static<typeof MixerDatum>;
export const MixerDatum = Data.Enum([
  Data.Object({
    Tree: Data.Tuple([MerkleTree]),
  }),
  Data.Object({
    Nullifiers: Data.Tuple([Data.Array(Data.Bytes())]),
  }),
  Data.Literal('Vault'),
]);

export type MixerRedeemer = Data.Static<typeof MixerRedeemer>;
export const MixerRedeemer = Data.Enum([
  Data.Object({
    Deposit: Data.Tuple([Data.Bytes()]),
  }),
  Data.Object({
    Withdraw: Data.Tuple([Data.Bytes()]),
  }),
]);

export type MintRedeemer = Data.Static<typeof MintRedeemer>;
export const MintRedeemer = Data.Array(Data.Bytes());
