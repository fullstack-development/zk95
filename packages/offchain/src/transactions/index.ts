export * from './deploy-pool';
export * from './deposit';
export * from './withdraw';
