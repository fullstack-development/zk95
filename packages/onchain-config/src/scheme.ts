import { Data } from 'lucid-cardano';

const MerkleTreeConfig = Data.Object({
  height: Data.Integer(),
  zeroHash: Data.Bytes(),
});

export const MixerConfig = Data.Object({
  protocolPolicyId: Data.Bytes(),
  protocolTokenName: Data.Bytes(),
  poolNominal: Data.Integer(),
  merkleTreeConfig: MerkleTreeConfig,
});
