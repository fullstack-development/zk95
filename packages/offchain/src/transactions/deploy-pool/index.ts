import { Lucid, Data, Assets, toHex, fromText } from 'lucid-cardano';
import { createMixerValidator } from './create-mixer-validator';
import { createMintingPolicy } from './create-minting-policy';

import { makeMerkleTree } from '@mixer/merkletree';
import { MintRedeemer, MixerDatum } from '../../scheme';
import { PoolInfo } from '../../types';
import { assert } from '@mixer/utils';

export async function deployPool(
  lucid: Lucid,
  nominal: bigint,
  treeHeight: number,
  zeroValue: string,
  treeTokenName: string,
  vaultTokenName: string,
  nullifiersTokenName: string
): Promise<PoolInfo> {
  const { script: mintingPolicyScript, policyId } = await createMintingPolicy(
    lucid
  );

  const ownAddress = await lucid.wallet.address();
  const paymentCredential =
    lucid.utils.getAddressDetails(ownAddress).paymentCredential;

  assert('Could not get payment key hash', paymentCredential?.type === 'Key');

  const { script: mixerValidator, address: mixerAddress } =
    createMixerValidator(
      lucid,
      policyId,
      fromText(treeTokenName),
      fromText(nullifiersTokenName),
      fromText(vaultTokenName),
      paymentCredential.hash,
      nominal,
      BigInt(treeHeight),
      zeroValue
    );

  const treeTokenUnit = policyId + fromText(treeTokenName);
  const vaultTokenUnit = policyId + fromText(vaultTokenName);
  const nullifiersTokenUnit = policyId + fromText(nullifiersTokenName);
  const mixerAsset: Assets = {
    [treeTokenUnit]: BigInt(1),
    [vaultTokenUnit]: BigInt(1),
    [nullifiersTokenUnit]: BigInt(1),
  };

  const emptyMerkleTree = makeMerkleTree({
    height: treeHeight,
    leafs: [],
    zeroValue,
  });

  const treeDatum = Data.to<MixerDatum>(
    {
      Tree: [
        {
          root: toHex(emptyMerkleTree.root.slice(1)),
          leafs: [],
        },
      ],
    },
    MixerDatum as never
  );

  const nullifiersDatum = Data.to<MixerDatum>(
    {
      Nullifiers: [[]],
    },
    MixerDatum as never
  );

  const vaultDatum = Data.to<MixerDatum>('Vault', MixerDatum as never);

  const mintRedeemer = Data.to<MintRedeemer>(
    [
      fromText(treeTokenName),
      fromText(vaultTokenName),
      fromText(nullifiersTokenName),
    ],
    MintRedeemer as never
  );

  const tx = await lucid
    .newTx()
    .attachMintingPolicy(mintingPolicyScript)
    .mintAssets(mixerAsset, mintRedeemer)
    .payToContract(
      mixerAddress,
      {
        inline: treeDatum,
      },
      {
        [treeTokenUnit]: BigInt(1),
        lovelace: BigInt(2000000),
      }
    )
    .payToContract(
      mixerAddress,
      {
        inline: nullifiersDatum,
      },
      {
        [nullifiersTokenUnit]: BigInt(1),
        lovelace: BigInt(2000000),
      }
    )
    .payToContract(
      mixerAddress,
      {
        inline: vaultDatum,
      },
      { [vaultTokenUnit]: BigInt(1) }
    )
    .payToAddressWithData(
      mixerAddress,
      {
        scriptRef: mixerValidator,
      },
      {}
    )
    .complete();

  const signedTx = await tx.sign().complete();

  const txHash = await signedTx.submit();

  return {
    txHash,
    network: lucid.network,
    nominal: nominal.toString(),
    treeHeight,
    treeTokenUnit,
    vaultTokenUnit,
    nullifiersTokenUnit,
    zeroValue,
    address: mixerAddress,
    script: mixerValidator,
  };
}
