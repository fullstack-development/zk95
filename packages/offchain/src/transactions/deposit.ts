import { Data, Lucid, addAssets, toHex } from 'lucid-cardano';
import { makeMerkleTree } from '@mixer/merkletree';
import { assert } from '@mixer/utils';
import { MixerDatum, MixerRedeemer } from '../scheme';
import { PoolInfo } from '../types';

export async function deposit(
  lucid: Lucid,
  {
    address,
    treeTokenUnit,
    vaultTokenUnit,
    nominal,
    treeHeight,
    zeroValue,
  }: Pick<
    PoolInfo,
    | 'address'
    | 'treeTokenUnit'
    | 'vaultTokenUnit'
    | 'nominal'
    | 'treeHeight'
    | 'zeroValue'
  >,
  commitmentHash: Uint8Array
) {
  const depositUTxOs = await lucid.utxosAt(address);
  const utxoWithScriptRef = depositUTxOs.find(
    ({ scriptRef }) =>
      scriptRef && lucid.utils.validatorToAddress(scriptRef) === address
  );

  assert(
    `Cannot find a utxo with the deposit reference script`,
    utxoWithScriptRef
  );

  const vaultUTxO = depositUTxOs.find(
    ({ assets }) => assets[vaultTokenUnit] === BigInt(1)
  );

  assert(`Cannot find a utxo with the vault`, vaultUTxO);
  assert(`Cannot find a vault datum`, vaultUTxO.datum);

  const merkleTreeUTxO = depositUTxOs.find(
    ({ assets }) => assets[treeTokenUnit] === BigInt(1)
  );

  assert(`Cannot find a utxo with the merkle tree`, merkleTreeUTxO);
  assert(`Merkle tree datum is missing`, merkleTreeUTxO.datum);

  const redeemer = Data.to<MixerRedeemer>(
    {
      Deposit: [toHex(commitmentHash)],
    },
    MixerRedeemer as never
  );

  const inputDatum = Data.from<MixerDatum>(
    merkleTreeUTxO.datum,
    MixerDatum as never
  );

  assert(
    `Wrong datum in the merkle tree utxo`,
    inputDatum !== 'Vault' && 'Tree' in inputDatum
  );

  const merkleTree = makeMerkleTree({
    height: treeHeight,
    leafs: inputDatum.Tree[0].leafs,
    zeroValue,
  });

  assert('Pool is full', !merkleTree.completed);

  const nextMerkleTree = merkleTree.insert(commitmentHash);

  const outputDatum = Data.to<MixerDatum>(
    {
      Tree: [
        {
          root: toHex(nextMerkleTree.root.slice(1)),
          leafs: nextMerkleTree.leafs.map(toHex),
        },
      ],
    },
    MixerDatum as never
  );

  const tx = await lucid
    .newTx()
    .readFrom([utxoWithScriptRef])
    .collectFrom([merkleTreeUTxO], redeemer)
    .payToContract(address, { inline: outputDatum }, merkleTreeUTxO.assets)
    .collectFrom([vaultUTxO], redeemer)
    .payToContract(
      address,
      { inline: vaultUTxO.datum },
      addAssets(vaultUTxO.assets, {
        lovelace: BigInt(nominal),
      })
    )
    .complete();

  const signedTx = await tx.sign().complete();
  const txHash = await signedTx.submit();
  return txHash;
}
