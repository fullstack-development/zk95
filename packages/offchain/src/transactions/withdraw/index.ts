import { Address, Data, Lucid, addAssets, toHex } from 'lucid-cardano';
import { assert } from '@mixer/utils';
import { MixerDatum, MixerRedeemer } from '../../scheme';
import { PoolInfo } from '../../types';

export async function withdraw(
  lucid: Lucid,
  {
    address,
    vaultTokenUnit,
    nullifiersTokenUnit,
    nominal,
  }: Pick<
    PoolInfo,
    'address' | 'vaultTokenUnit' | 'nullifiersTokenUnit' | 'nominal'
  >,
  recipient: Address,
  feeLovelace: bigint,
  nullifier: Uint8Array
) {
  const depositUTxOs = await lucid.utxosAt(address);

  const utxoWithScriptRef = depositUTxOs.find(
    ({ scriptRef }) =>
      scriptRef && lucid.utils.validatorToAddress(scriptRef) === address
  );

  assert(
    `Cannot find a utxo with the deposit reference script`,
    utxoWithScriptRef
  );

  const vaultUTxO = depositUTxOs.find(
    ({ assets }) => assets[vaultTokenUnit] === BigInt(1)
  );

  assert(`Cannot find a utxo with the vault`, vaultUTxO);
  assert(`Cannot find a vault datum`, vaultUTxO.datum);

  const nullifiersUTxO = depositUTxOs.find(
    ({ assets }) => assets[nullifiersTokenUnit] === BigInt(1)
  );

  assert(`Cannot find a utxo with the nullifiers`, nullifiersUTxO);
  assert(`Nullifiers datum is missing`, nullifiersUTxO.datum);

  const redeemer = Data.to<MixerRedeemer>(
    {
      Withdraw: [toHex(nullifier)],
    },
    MixerRedeemer as never
  );

  const inputDatum = Data.from<MixerDatum>(
    nullifiersUTxO.datum,
    MixerDatum as never
  );

  assert(`Wrong datum`, inputDatum !== 'Vault' && 'Nullifiers' in inputDatum);

  const outputDatum = Data.to<MixerDatum>(
    {
      Nullifiers: [inputDatum.Nullifiers[0].concat([toHex(nullifier)])],
    },
    MixerDatum as never
  );

  const ownAddress = await lucid.wallet.address();

  const tx = await lucid
    .newTx()
    .readFrom([utxoWithScriptRef])
    .collectFrom([nullifiersUTxO], redeemer)
    .payToContract(address, { inline: outputDatum }, nullifiersUTxO.assets)
    .collectFrom([vaultUTxO], redeemer)
    .payToContract(
      address,
      { inline: vaultUTxO.datum },
      addAssets(vaultUTxO.assets, {
        lovelace: -BigInt(nominal),
      })
    )
    .payToAddress(recipient, {
      lovelace: BigInt(nominal) - feeLovelace,
    })
    .addSigner(ownAddress)
    .complete();

  const signedTx = await tx.sign().complete();
  const txHash = await signedTx.submit();
  return txHash;
}
