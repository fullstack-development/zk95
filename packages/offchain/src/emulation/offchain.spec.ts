import {
  Lucid,
  Emulator,
  fromUnit,
  toText,
  Data,
  generatePrivateKey,
  Provider,
  generateSeedPhrase,
} from 'lucid-cardano';
import { jest } from '@jest/globals';

import { hash, hashConcat } from '@mixer/crypto';

import { deployPool, deposit, withdraw } from '../transactions';
import { MixerDatum } from '../scheme';

const mainMnemonicPhrase =
  'edge shadow topple brush online kid quit north muffin donate accident endorse other grant sleep';
const recipientMnemonicPhrase =
  'nut borrow depart parade aerobic pause equal produce secret avocado wild clump advance just safe slot cat forward luggage brother surround drive goose book';
const treeHeight = 7;
const zeroValue = 'tornado.cash on cardano';

describe('offchain', () => {
  it('deploy pool', async () => {
    const [lucid, emulator] = await prepareEmulator();
    await deployPool(
      lucid,
      BigInt(100000000),
      treeHeight,
      zeroValue,
      'Tree',
      'Vault',
      'Nullifiers'
    );
    emulator.awaitBlock(1);

    expect(printLedger(emulator.ledger)).toMatchSnapshot('deployment ledger');
  });

  it('deposit', async () => {
    const [lucid, emulator] = await prepareEmulator();
    const poolInfo = await deployPool(
      lucid,
      BigInt(100000000),
      treeHeight,
      zeroValue,
      'Tree',
      'Vault',
      'Nullifiers'
    );
    emulator.awaitBlock(1);

    const commitmentHash = hash('commitment');
    await deposit(lucid, poolInfo, commitmentHash);

    emulator.awaitBlock(1);

    expect(printLedger(emulator.ledger)).toMatchSnapshot('deposit ledger');
  });

  it('double deposit', async () => {
    const [lucid, emulator] = await prepareEmulator();
    const poolInfo = await deployPool(
      lucid,
      BigInt(100000000),
      treeHeight,
      zeroValue,
      'Tree',
      'Vault',
      'Nullifiers'
    );
    emulator.awaitBlock(1);

    await deposit(lucid, poolInfo, hash('commitment1'));

    emulator.awaitBlock(1);

    await deposit(lucid, poolInfo, hash('commitment2'));

    emulator.awaitBlock(1);

    expect(printLedger(emulator.ledger)).toMatchSnapshot(
      'double deposit ledger'
    );
  });

  it('withdraw', async () => {
    const [lucid, emulator] = await prepareEmulator();
    const poolInfo = await deployPool(
      lucid,
      BigInt(100000000),
      treeHeight,
      zeroValue,
      'Tree',
      'Vault',
      'Nullifiers'
    );
    emulator.awaitBlock(1);
    const nullifier1 = hash('nullifier 1').slice(0, 31);
    const secret1 = hash('secret 1').slice(0, 31);
    const commitmentHash1 = hashConcat(nullifier1, secret1);
    await deposit(lucid, poolInfo, commitmentHash1);
    emulator.awaitBlock(1);

    const recipientAddress = await getAddress(recipientMnemonicPhrase);
    await withdraw(
      lucid,
      poolInfo,
      recipientAddress,
      BigInt(2000000),
      nullifier1
    );
    emulator.awaitBlock(1);

    expect(printLedger(emulator.ledger)).toMatchSnapshot('withdraw ledger');
  });

  it('double withdraw', async () => {
    const [lucid, emulator] = await prepareEmulator();
    const poolInfo = await deployPool(
      lucid,
      BigInt(100000000),
      treeHeight,
      zeroValue,
      'Tree',
      'Vault',
      'Nullifiers'
    );
    emulator.awaitBlock(1);
    const nullifier1 = hash('nullifier 1').slice(0, 31);
    const secret1 = hash('secret 1').slice(0, 31);
    const commitmentHash1 = hashConcat(nullifier1, secret1);
    await deposit(lucid, poolInfo, commitmentHash1);
    emulator.awaitBlock(1);

    const recipientAddress = await getAddress(recipientMnemonicPhrase);
    await withdraw(
      lucid,
      poolInfo,
      recipientAddress,
      BigInt(2000000),
      nullifier1
    );
    emulator.awaitBlock(1);

    const spy = jest.fn();
    try {
      await withdraw(
        lucid,
        poolInfo,
        recipientAddress,
        BigInt(2000000),
        nullifier1
      );
      emulator.awaitBlock(1);
    } catch (error) {
      spy(error);
    }

    expect(spy).toBeCalled();
  });

  it('withdraw with wrong signature', async () => {
    const [lucid, emulator] = await prepareEmulator();
    const poolInfo = await deployPool(
      lucid,
      BigInt(100000000),
      treeHeight,
      zeroValue,
      'Tree',
      'Vault',
      'Nullifiers'
    );
    emulator.awaitBlock(1);

    const nullifier1 = hash('nullifier 1').slice(0, 31);
    const secret1 = hash('secret 1').slice(0, 31);
    const commitmentHash1 = hashConcat(nullifier1, secret1);
    await deposit(lucid, poolInfo, commitmentHash1);
    emulator.awaitBlock(1);

    const spy = jest.fn();
    try {
      const fakeLucid = await mkLucid();
      await withdraw(
        fakeLucid,
        poolInfo,
        await fakeLucid.wallet.address(),
        BigInt(2000000),
        nullifier1
      );
      emulator.awaitBlock(1);
    } catch (error) {
      spy(error);
    }

    expect(spy).toBeCalled();
  });
});

async function prepareEmulator(): Promise<[Lucid, Emulator]> {
  const address = await getAddress(mainMnemonicPhrase);
  const emulator = new Emulator([
    { address, assets: { lovelace: 3000000000n } },
  ]);

  const lucid = await mkLucid(mainMnemonicPhrase, emulator);
  return [lucid, emulator];
}

async function mkLucid(seed?: string, provider?: Provider) {
  const lucid = await Lucid.new(provider, 'Custom');
  return seed
    ? lucid.selectWalletFromSeed(seed)
    : lucid.selectWalletFromPrivateKey(generatePrivateKey());
}

async function getAddress(seed?: string) {
  const lucid = await mkLucid(seed);
  return lucid.wallet.address();
}

function printLedger(ledger: Emulator['ledger']) {
  return Object.keys(ledger).reduce((newLedger, txHash) => {
    const { address, assets, datum, datumHash, scriptRef, outputIndex } =
      ledger[txHash].utxo;

    const newAssets = Object.keys(assets).reduce((newAssets, unit) => {
      const assetDetails = fromUnit(unit);

      newAssets[assetDetails.policyId] = {
        tokenName: assetDetails.name ? toText(assetDetails.name) : null,
        quantity: assets[unit],
      };

      return newAssets;
    }, {} as any);

    const parsedDatum = datum
      ? (() => {
          try {
            return Data.from(datum, MixerDatum as never);
          } catch (error) {
            return datum;
          }
        })()
      : undefined;

    newLedger[txHash] = {
      address,
      assets: newAssets,
      datum: parsedDatum,
      datumHash,
      hasScriptRef: Boolean(scriptRef),
      outputIndex,
    };

    return newLedger;
  }, {} as any);
}
