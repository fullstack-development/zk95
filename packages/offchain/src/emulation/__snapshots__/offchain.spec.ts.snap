// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offchain deploy pool: deployment ledger 1`] = `
{
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c340": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Tree",
      },
      "lovelace": {
        "quantity": 2000000n,
        "tokenName": null,
      },
    },
    "datum": {
      "Tree": [
        {
          "leafs": [],
          "root": "6e4171d3e3e1c7f39a7b4b0d94509bdd3e3ef44ce3c0d2f3e9d43aaf160b28",
        },
      ],
    },
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 0,
  },
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c341": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Nullifiers",
      },
      "lovelace": {
        "quantity": 2000000n,
        "tokenName": null,
      },
    },
    "datum": {
      "Nullifiers": [
        [],
      ],
    },
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 1,
  },
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c342": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Vault",
      },
      "lovelace": {
        "quantity": 1068880n,
        "tokenName": null,
      },
    },
    "datum": "Vault",
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 2,
  },
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c343": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "lovelace": {
        "quantity": 24067040n,
        "tokenName": null,
      },
    },
    "datum": undefined,
    "datumHash": undefined,
    "hasScriptRef": true,
    "outputIndex": 3,
  },
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c344": {
    "address": "addr_test1qpuks2fj3l0umdy8h828zkz25sslwnrtzs7weed04lma8dkj7fmpe62uxcuwe5anz4kks022pj6fxa0mcfz52t9a9vysx0qvdu",
    "assets": {
      "lovelace": {
        "quantity": 2970396466n,
        "tokenName": null,
      },
    },
    "datum": undefined,
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 4,
  },
}
`;

exports[`offchain deposit: deposit ledger 1`] = `
{
  "478298643c4a3abb3d2717ee3e78f38d02dc0262f7740a87f508691d625adea90": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Tree",
      },
      "lovelace": {
        "quantity": 2000000n,
        "tokenName": null,
      },
    },
    "datum": {
      "Tree": [
        {
          "leafs": [
            "6a19f0fb4be54511524bcd5b0c98b38da1ee049a39735c39311e10336024436f",
          ],
          "root": "2ff91aa4bc7c37075ff1c4911fe6fef7a3df5496cdb5e1767b1998db3c7489",
        },
      ],
    },
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 0,
  },
  "478298643c4a3abb3d2717ee3e78f38d02dc0262f7740a87f508691d625adea91": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Vault",
      },
      "lovelace": {
        "quantity": 101068880n,
        "tokenName": null,
      },
    },
    "datum": "Vault",
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 1,
  },
  "478298643c4a3abb3d2717ee3e78f38d02dc0262f7740a87f508691d625adea92": {
    "address": "addr_test1qpuks2fj3l0umdy8h828zkz25sslwnrtzs7weed04lma8dkj7fmpe62uxcuwe5anz4kks022pj6fxa0mcfz52t9a9vysx0qvdu",
    "assets": {
      "lovelace": {
        "quantity": 2870130410n,
        "tokenName": null,
      },
    },
    "datum": undefined,
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 2,
  },
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c341": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Nullifiers",
      },
      "lovelace": {
        "quantity": 2000000n,
        "tokenName": null,
      },
    },
    "datum": {
      "Nullifiers": [
        [],
      ],
    },
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 1,
  },
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c343": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "lovelace": {
        "quantity": 24067040n,
        "tokenName": null,
      },
    },
    "datum": undefined,
    "datumHash": undefined,
    "hasScriptRef": true,
    "outputIndex": 3,
  },
}
`;

exports[`offchain double deposit: double deposit ledger 1`] = `
{
  "0b9d09dfb9f5e8b968e2632f1da37c247b3598a5b5c73a1ae32ad169d2b89cd10": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Tree",
      },
      "lovelace": {
        "quantity": 2000000n,
        "tokenName": null,
      },
    },
    "datum": {
      "Tree": [
        {
          "leafs": [
            "e1587401f42aa18899004e6b313f62b1732dd9860ac4085f4790f62b24b7dec3",
            "bdde83d5b26cd7677153640af3ce94739df8d51443024f2933abf843b8c1b009",
          ],
          "root": "1c4b51b53c4ed72deafb6f3e9ebca853bc9f6db6828fad536d20d1fa84b3f1",
        },
      ],
    },
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 0,
  },
  "0b9d09dfb9f5e8b968e2632f1da37c247b3598a5b5c73a1ae32ad169d2b89cd11": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Vault",
      },
      "lovelace": {
        "quantity": 201068880n,
        "tokenName": null,
      },
    },
    "datum": "Vault",
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 1,
  },
  "0b9d09dfb9f5e8b968e2632f1da37c247b3598a5b5c73a1ae32ad169d2b89cd12": {
    "address": "addr_test1qpuks2fj3l0umdy8h828zkz25sslwnrtzs7weed04lma8dkj7fmpe62uxcuwe5anz4kks022pj6fxa0mcfz52t9a9vysx0qvdu",
    "assets": {
      "lovelace": {
        "quantity": 2769861218n,
        "tokenName": null,
      },
    },
    "datum": undefined,
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 2,
  },
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c341": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Nullifiers",
      },
      "lovelace": {
        "quantity": 2000000n,
        "tokenName": null,
      },
    },
    "datum": {
      "Nullifiers": [
        [],
      ],
    },
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 1,
  },
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c343": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "lovelace": {
        "quantity": 24067040n,
        "tokenName": null,
      },
    },
    "datum": undefined,
    "datumHash": undefined,
    "hasScriptRef": true,
    "outputIndex": 3,
  },
}
`;

exports[`offchain withdraw: withdraw ledger 1`] = `
{
  "0110cfe24e5c84ad373443ba6f25ecb831912fbd14f99d5e80228d17ae5ea3d90": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Nullifiers",
      },
      "lovelace": {
        "quantity": 2000000n,
        "tokenName": null,
      },
    },
    "datum": {
      "Nullifiers": [
        [
          "9d59ae5baedd7006e63b61fd79f83ceb5c71d02433bcc95ca12939b3344461",
        ],
      ],
    },
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 0,
  },
  "0110cfe24e5c84ad373443ba6f25ecb831912fbd14f99d5e80228d17ae5ea3d91": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Vault",
      },
      "lovelace": {
        "quantity": 1068880n,
        "tokenName": null,
      },
    },
    "datum": "Vault",
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 1,
  },
  "0110cfe24e5c84ad373443ba6f25ecb831912fbd14f99d5e80228d17ae5ea3d92": {
    "address": "addr_test1qz4m06a7vfrlxecwz9cq5d85htr0h8qzkglse39e02xd790q0lh95wfaeym84qw5th90mhw0uanx2sxdxzpwhl9eq2csk66908",
    "assets": {
      "lovelace": {
        "quantity": 98000000n,
        "tokenName": null,
      },
    },
    "datum": undefined,
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 2,
  },
  "0110cfe24e5c84ad373443ba6f25ecb831912fbd14f99d5e80228d17ae5ea3d93": {
    "address": "addr_test1qpuks2fj3l0umdy8h828zkz25sslwnrtzs7weed04lma8dkj7fmpe62uxcuwe5anz4kks022pj6fxa0mcfz52t9a9vysx0qvdu",
    "assets": {
      "lovelace": {
        "quantity": 2871871885n,
        "tokenName": null,
      },
    },
    "datum": undefined,
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 3,
  },
  "8e14b76f36d8ce0ac065403907749f18f31ce6c70bb2ec7c74178d58192d6f560": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "516feef38e451685a9043aa91b26191f8d4c1cb20c6d93f00e9c8368": {
        "quantity": 1n,
        "tokenName": "Tree",
      },
      "lovelace": {
        "quantity": 2000000n,
        "tokenName": null,
      },
    },
    "datum": {
      "Tree": [
        {
          "leafs": [
            "e9e19a86818cb673ece879e6727cfaaa9495a08337738f697025ea5263e847c4",
          ],
          "root": "80e7f99be3734e72d5e68710b0c35573db6c08ce21166d47cce9dbbe544353",
        },
      ],
    },
    "datumHash": undefined,
    "hasScriptRef": false,
    "outputIndex": 0,
  },
  "a20bd421af51eb92cb3000fa8423e9d5124ec238583e39cdafec944b62750c343": {
    "address": "addr_test1wrfygkz4zfg466ne68mezx240zj28098l33zxqld2s2u79slxrwq0",
    "assets": {
      "lovelace": {
        "quantity": 24067040n,
        "tokenName": null,
      },
    },
    "datum": undefined,
    "datumHash": undefined,
    "hasScriptRef": true,
    "outputIndex": 3,
  },
}
`;
